catalog/resource_catalog.go

```diff
@@ -80,39 +41,79 @@ func ResourceCatalog() *schema.Resource {
 			m["storage_root"].DiffSuppressFunc = ucDirectoryPathSuppressDiff
 			return m
 		})
-	update := updateFunctionFactory("/unity-catalog/catalogs", []string{"owner", "comment", "properties"})
 	return common.Resource{
 		Schema: catalogSchema,
 		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
-			var ci CatalogInfo
-			common.DataToStructPointer(d, catalogSchema, &ci)
-			if err := NewCatalogsAPI(ctx, c).createCatalog(&ci); err != nil {
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+
+			var createCatalogRequest catalog.CreateCatalog
+			common.DataToStructPointer(d, catalogSchema, &createCatalogRequest)
+			ci, err := w.Catalogs.Create(ctx, createCatalogRequest)
+			if err != nil {
 				return err
 			}
 			// only remove catalog default schema for non-Delta Sharing catalog
 			if ci.ShareName == "" {
-				if err := NewSchemasAPI(ctx, c).deleteSchema(ci.Name + ".default"); err != nil {
+				if err := w.Schemas.DeleteByFullName(ctx, ci.Name+".default"); err != nil {
 					return fmt.Errorf("cannot remove new catalog default schema: %w", err)
 				}
 			}
+
 			d.SetId(ci.Name)
-			return update(ctx, d, c)
+
+			// Update owner or isolation mode if it is provided
+			if d.Get("owner") == "" && d.Get("isolation_mode") == "" {
+				return nil
+			}
+			var updateCatalogRequest catalog.UpdateCatalog
+			common.DataToStructPointer(d, catalogSchema, &updateCatalogRequest)
+			_, err = w.Catalogs.Update(ctx, updateCatalogRequest)
+			if err != nil {
+				return err
+			}
+
+			return nil
 		},
 		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
-			ci, err := NewCatalogsAPI(ctx, c).getCatalog(d.Id())
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+
+			ci, err := w.Catalogs.GetByName(ctx, d.Id())
 			if err != nil {
 				return err
 			}
 			return common.StructToData(ci, catalogSchema, d)
 		},
-		Update: update,
+		Update: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+			var updateCatalogRequest catalog.UpdateCatalog
+			common.DataToStructPointer(d, catalogSchema, &updateCatalogRequest)
+			ci, err := w.Catalogs.Update(ctx, updateCatalogRequest)
```

Comment:
How it will work when the name is already updated in resource - as I see in the docs, we address catalog by name: https://docs.databricks.com/api/workspace/catalogs/update, so if the name in resource is updated, it won't find an old one to change it.

Reply:
if the change happened out of bound, then there is not much we can do. what happens with the update operation here is that we call the Update API, update the resource Id to the new name, so that the Get API will use the new name and succeed

Reply:
credit to @tanmay-db for getting this to worked with the Volume resource, so I am just backporting the fix

Reply:
If we allow name change, then SDK needs to support that pattern. Inside provider we can find if the name field has changed, and then use previous name (the `d.Id()`) in the URL

Reply:
we need to do the same for all the UC resources then, as they are all addressed by names.

currently volume handles it like this https://github.com/databricks/terraform-provider-databricks/blob/master/catalog/resource_volume.go#L85